src/ecmult_multi_impl.h

```diff
@@ -0,0 +1,252 @@
+             * reduces the magnitude of the larger scalar, on average by half. So by
+             * repeating this we will quickly zero out all but one exponent, which will
+             * be small. */
+            secp256k1_gej_add_var(&pt[second], &pt[first], &pt[second], NULL);  /* Y -> X + Y */
+            secp256k1_scalar_numsub(&sc[first], &sc[first], &sc[second]);  /* n -> n - m */
+
+            if (secp256k1_scalar_cmp_var(&sc[first], &sc[second]) < 0) {
+                break;
+            }
+
+            /* To handle pathological inputs, we use a binary ladder step here */
+            if (secp256k1_scalar_shr_int(&sc[first], 1) == 1) {
+                secp256k1_gej_add_var(r, r, &pt[first], NULL);
+            }
+            secp256k1_gej_double_var(&pt[first], &pt[first], NULL);
+        }
+        while (secp256k1_scalar_cmp_var(&sc[first], &sc[second]) >= 0);
+
+        if (secp256k1_scalar_is_zero(&sc[first])) {
+            first = secp256k1_heap_remove(&heap);
+        } else {
+            first = secp256k1_replace(&heap, first);
+        }
+    }
+
+    VERIFY_CHECK(!secp256k1_scalar_is_zero(&sc[first]));
+
+    /* Now the desired result is heap_sc[0] * heap_pt[0], and for random scalars it is
+     * very likely that heap_sc[0] = 1, and extremely likely heap_sc[0] < 5. (After
+     * about 100k trials I saw around 200 2's and one 3.) So use a binary ladder rather
+     * than any heavy machinery to finish it off. */
+    for (;;) {
+        if (secp256k1_scalar_shr_int(&sc[first], 1) == 1) {
+            secp256k1_gej_add_var(r, r, &pt[first], NULL);
+            if (secp256k1_scalar_is_zero(&sc[first])) {
+                break;
+            }
+        }
+        secp256k1_gej_double_var(&pt[first], &pt[first], NULL);
+    }
+}
+
+#ifdef USE_ENDOMORPHISM
+SECP256K1_INLINE static void secp256k1_ecmult_endo_split(secp256k1_scalar *s1, secp256k1_scalar *s2, secp256k1_gej *p1, secp256k1_gej *p2) {
+    secp256k1_scalar tmp = *s1;
+    secp256k1_scalar_split_lambda(s1, s2, &tmp);
+    secp256k1_gej_mul_lambda(p2, p1);
+
+    if (secp256k1_scalar_is_high(s1)) {
+        secp256k1_scalar_negate(s1, s1);
+        secp256k1_gej_neg(p1, p1);
+    }
+    if (secp256k1_scalar_is_high(s2)) {
+        secp256k1_scalar_negate(s2, s2);
+        secp256k1_gej_neg(p2, p2);
+    }
+}
+#endif
+
+static int secp256k1_ecmult_multi(secp256k1_gej *r, const secp256k1_scalar *inp_g_sc, secp256k1_ecmult_multi_callback cb, void *cbdata, size_t n) {
+    secp256k1_gej tmp;
+    secp256k1_gej pt[SECP256K1_ECMULT_MULTI_MAX_N + 1];  /* +1 in case we spill over doing the endomorphism 2 points at a time */
+    secp256k1_scalar sc[SECP256K1_ECMULT_MULTI_MAX_N + 1];
+    size_t idx = 0;
+    size_t point_idx = 0;
+
+    sc[0] = *inp_g_sc;
+    secp256k1_gej_set_ge(&pt[0], &secp256k1_ge_const_g);
+    idx++;
+#ifdef USE_ENDOMORPHISM
+    secp256k1_ecmult_endo_split(&sc[0], &sc[1], &pt[0], &pt[1]);
```

Comment:
Doing endo split on the G factor is unnecessary... you can just split in 128-bit parts, and a precomputed 2^128*G constant (which we probably already have in the verification context).